extensions/amp-a4a/0.1/signature-verifier.js

```diff
@@ -0,0 +1,357 @@
+                          } else {
+                            return VerificationFailure.SIGNATURE_MISMATCH;
+                          }
+                        },
+                        err => {
+                          // Web Cryptography rejected the verification attempt.
+                          // This hopefully won't happen in the wild, but
+                          // browsers can be weird about this, so we need to
+                          // guard against the possibility. Phone home to the
+                          // AMP Project so that we can understand why this
+                          // occurred.
+                          const message = err && err.message;
+                          dev().error(
+                              'AMP-A4A',
+                              `Failed to verify signature: ${message}`);
+                          return VerificationFailure.NO_FAULT;
+                        });
+              } else {
+                // This particular public key couldn't be imported. Probably the
+                // signing service's fault.
+                return VerificationFailure.NO_FAULT;
+              }
+            });
+          }
+        } else {
+          // The public keyset couldn't be fetched and imported. Probably a
+          // network connectivity failure.
+          return VerificationFailure.NO_FAULT;
+        }
+      });
+    } else {
+      // Web Cryptography isn't available.
+      return Promise.resolve(
+          /** @type {?VerificationFailure} */ (VerificationFailure.NO_FAULT));
+    }
+  }
+
+  /**
+   * Try to download the keyset for the named signing service and add a promise
+   * for each key to the `keys` object.
+   *
+   * @param {!Object<string, ?Promise<?webCrypto.CryptoKey>>} keys the object to
+   *     add each key promise to. This is mutated while the returned promise is
+   *     pending.
+   * @param {string} signingServiceName
+   * @param {?string} keypairId the keypair ID to include in the query string
+   *     for cachebusting purposes, or `null` if no cachebusting is needed
+   * @return {!Promise<boolean>} resolves after the mutation of `keys` is
+   *     complete, to `true` if the keyset was downloaded and parsed
+   *     successfully (even if some keys were malformed), or `false` if a
+   *     keyset-level failure occurred
+   * @private
+   */
+  fetchAndAddKeys_(keys, signingServiceName, keypairId) {
+    let url = signingServerURLs[signingServiceName];
+    if (keypairId != null) {
+      url += '?' + encodeURIComponent(keypairId);
+    }
+    return xhrFor(this.win_)
+        .fetch(
+            url,
+            {mode: 'cors', method: 'GET', ampCors: false, credentials: 'omit'})
+        .then(
+            response => {
+              if (response.status == 200) {
+                // The signing service spec requires this, but checking it at
+                // runtime in production isn't worth it.
+                dev().assert(
```

Comment:
Why only dev?  This could definitely occur in production and so we should guard against it, no?